memclear.c

```diff
@@ -90,30 +110,125 @@ EFI_STATUS efi_main(EFI_HANDLE image_handle, EFI_SYSTEM_TABLE* system_table)
+			//debug_hex(L"Cleaned the conventional memory region: 0x", start);
+		}
+		else {
+
+			int Clearable = 0;
+			switch (Map->Type) {
+				// Becareful: The following mapped regions will be cleared
+				case EfiACPIReclaimMemory:
+				case EfiACPIMemoryNVS:
+				case EfiMemoryMappedIO:
+				case EfiMemoryMappedIOPortSpace:
+					Clearable = 1;
+					break;
+				case EfiBootServicesData: 
+					if (rsp < start || rsp > end) {
+						Clearable = 1;
+						break;
+					}
+				case EfiBootServicesCode:
+					if (rip < start || rip > end) {
+						Clearable = 1;
+						break;
+					}
+
+					// The following regions will not be cleared but locked out
+					// through grub
+				case EfiPalCode:
+				case EfiLoaderCode:
+				case EfiUnusableMemory:
+				case EfiReservedMemoryType:
+				case EfiRuntimeServicesCode:
+				case EfiRuntimeServicesData:
+					Clearable = 0;
+					break;
+			}
+
+
+			if (Clearable) {
+				CleanedMaps[NumCleanedMap].Start = start;
+				CleanedMaps[NumCleanedMap].End = end;
+				++NumCleanedMap;
+				debug_hex(L"Clearable ", start);
+			}
+			else {
+				debug_hex(L"Uncleaned ", start);
+				ProtectedMemSize += (end - start);
+			}
 		}
 	}
+	debug_hex(L"The total size of protectred memory: ", ProtectedMemSize);
 
 	/* Turn on interrupts again */
 	asm volatile("sti");
 
-	gErr->OutputString(gErr, L"Successfully cleared memory\n");
 
 	/* Done */
 	r = gBS->FreePool(MemoryMap);
 	if (r != EFI_SUCCESS)
 		goto err;
 
-	/* Need to return an error so the next boot entry loads */
-	return EFI_BUFFER_TOO_SMALL;
+	// 
+	// Record which regions will not be cleared here and expect
+	// "memmap" to do so through the OS command line
+	//
+
+#if 0
+	// MyData is the data you want to store
+	UINT8 MyData[] = {1,2,3,4,5,6};
+
+	UINTN DataSize = sizeof(MyData);
+	r = system_table->RuntimeServices->SetVariable(
+			VariableName,
+			&MyVariableGuid,
+			EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS,
+			DataSize,
+			MyData
+			);
+	if (EFI_ERROR(r)) {
+		debug_hex(L"Failed to set variable ", r);
+		goto err;
+	}
+	else {
+		debug_hex(L"Set the variable ", r);
+	}
+#endif
+
+#if 0
+	//
+	// Assuming no changes to the memory map, call ExitBootServices
+	//
+	r = gBS->ExitBootServices(image_handle, MapKey);
+	if (EFI_ERROR(r)) {
+		debug_hex(L"Failed to exit services ", r);
+		goto err;
+	}
+#endif
+
+	//
+	// IMPORTANT: Do not use any other vairable besides the ones on
+	// the stack because all other data will be cleared
+	// 
+	// The following would clear all reclaimable maps except for the
+	// stack and the code of the current EFI application
+	for (int i = 0; i < NumCleanedMap; ++i) {
+		gBS->SetMem((void *)CleanedMaps[NumCleanedMap].Start,
+				CleanedMaps[NumCleanedMap].End - CleanedMaps[NumCleanedMap].Start, 0);
+		debug_hex(L"Cleaned ", CleanedMaps[NumCleanedMap].Start);
+	}
+	gErr->OutputString(gErr, L"Successfully cleared memory\n");
+
+
+	return EFI_SUCCESS;
```

Comment:
With EFI_SUCCESS, you can no longer put memclear into a BootOrder variable before the actual payload you want to run.